### ISSUE TITLE
Random jython exceptions in Fiji 2.0.0-rc-59/1.51j and jython-shaded-2.7.0

### ISSUE THREAD

**Issue by rejsmont** [fiji/fiji#160]
Hi,

I have implemented a jython plugin some time ago. It used to work fine last year (latest commit on Nov 5, 2015). Now I had to use it again and I am running into pretty random java exceptions like the one below:

```
Exception in thread "AWT-EventQueue-0" Traceback (most recent call last):
  File "/Users/radoslaw.ejsmont/src/Fiji/FijiScripts/mColoc3D.py", line 541, in getListCellRendererComponent
    c = DefaultListCellRenderer.getListCellRendererComponent(
NameError: global name 'DefaultListCellRenderer' is not defined
Exception in thread "AWT-EventQueue-0" Traceback (most recent call last):
  File "/Users/radoslaw.ejsmont/src/Fiji/FijiScripts/mColoc3D.py", line 396, in updateSlice
    self.updateSlice3D(self.imp)
  File "/Users/radoslaw.ejsmont/src/Fiji/FijiScripts/mColoc3D.py", line 408, in updateSlice3D
    roi = ShapeRoi(impRoi, position = index)
NameError: global name 'ShapeRoi' is not defined
Exception in thread "AWT-EventQueue-0" Traceback (most recent call last):
  File "/Users/radoslaw.ejsmont/src/Fiji/FijiScripts/mColoc3D.py", line 541, in getListCellRendererComponent
    c = DefaultListCellRenderer.getListCellRendererComponent(
NameError: global name 'DefaultListCellRenderer' is not defined
```

After one of such exceptions is triggered, the jython import system goes haywire and fails to find even as basic classes as ImagePlus, etc. Naturally I have all required imports in place, but somehow jython interpreter looses them. I do not see any jython error or java exception before these import failures are triggered.

You can find code for my script at:
[https://github.com/HassanLab/FijiScripts/blob/master/mColoc3D.py](https://github.com/HassanLab/FijiScripts/blob/master/mColoc3D.py)

Thanks in advance for looking into it!

Best,

Radek

**Comment by ctrueden** [comment]
Sorry for the delay in reply, @rejsmont. This is a known issue when your Python code registers callbacks which survive the execution of the script itself.

There is some rather aggressive reference cleaning logic which tries to null things out as soon as the script execution completes. Without such logic, Jython scripts end up with huge memory leaks, since references within the script are retained permanently.

A while back, I briefly [played around with removing the reference cleaning logic](https://github.com/scijava/scripting-jython/compare/ref-cleaning) but then the memory leaks are back.

So in short: you have stumbled upon a mammoth bug in the SciJava Jython scripting support, for which I do not currently have a good solution. Sorry.

**Comment by rejsmont** [comment]
@ctrueden thanks for the update... and sorry to hear there's no easy solution :(

**Comment by rejsmont** [comment]
Today's XKCD reminded me of this bug :)

![Existential Bug Reports](https://imgs.xkcd.com/comics/existential_bug_reports.png "ISSUE: If we wait long enough, the Earth will eventually be consumed by the Sun. WORKAROUND: None.")

**Comment by ctrueden** [comment]
@rejsmont Sorry, I know it's frustrating.

I have an idea for working around it. It's not pretty, but it would give you the functionality you need, at least locally.

We could add a toggle for the reference cleaning, via a system property like `scijava.jython.skipRefCleaning`. If you set this system property to true, then the reference cleaner is disabled.

Once the bug is fixed properly, the system property can become a no-op. But for now, it would let your scripts work as intended, with the caveat that you would eventually need to restart Fiji due to insufficient memory.

What do you think?

**Comment by rejsmont** [comment]
@ctrueden that sounds like a good plan! Thanks for that!

**Comment by ctrueden** [comment]
@rejsmont Looking at the code carefully, there is no way the reference cleaning works as intended anymore on the current master branch. I am investigating the best way forward. It would be helpful to have a minimal example for testing—unfortunately, I cannot remember how I tested it before.

**Comment by ctrueden** [comment]
@rejsmont Good news! In my tests now, it seems like the memory leak is gone, with the switchover to the stock Jython JSR-223 script engine (which was done on 2017-03-14, and is not yet released). I used the following script:
```
# @IOService io
data = io.open("/Users/curtis/data/dub.tif")
print("data = " + str(data))
```
Where `dub.tif` is a 1GB 4D image.

Memory usage, as expected, spikes to >1200MB. But after a few second, performing a garbage collection operation drops memory back down to <50MB again. So, I guess the problem solved itself. Sorry for all the brouhaha.

Please test it yourself:

* Delete or move away `jars/scripting-jython-0.4.0.jar`.
* Download the [latest scripting-jython snapshot](http://maven.imagej.net/service/local/artifact/maven/redirect?r=snapshots&g=org.scijava&a=scripting-jython&v=0.4.1-SNAPSHOT&e=jar).
* Place the downloaded JAR in your `jars` folder.

Start up Fiji and try your script. Check whether A) the callbacks reliably work now; and B) whether memory gets cleaned up properly after script execution is fully complete.

**Comment by ctrueden** [comment]
Thanks for testing, @rejsmont. I will try to cut a new release (of `scripting-jython` as well as many other components needing release) before week's end.
